Business setup for onboarding wizard

claudiulodro opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes #54 .

This PR implements an initial onboarding wizard for setting up general site settings. This first version handles WooCommerce address info and Stripe settings. If we like the toggle for the Stripe settings, we'll need to add a Muriel toggle and migrate the core toggle used here to a Muriel one in the future.

Screenshots in details
<details>

<img width="579" alt="Screen Shot 2019-06-03 at 11 36 59 AM" src="https://user-images.githubusercontent.com/7317227/58827456-eceecc80-85f7-11e9-9332-09ea61aa6f84.png">
<img width="564" alt="Screen Shot 2019-06-03 at 12 12 19 PM" src="https://user-images.githubusercontent.com/7317227/58827847-e3199900-85f8-11e9-8ef6-a12b2ee1da06.png">
<img width="569" alt="Screen Shot 2019-06-03 at 12 14 16 PM" src="https://user-images.githubusercontent.com/7317227/58827978-34298d00-85f9-11e9-8143-f7165de16a65.png">
<img width="558" alt="Screen Shot 2019-06-03 at 11 37 07 AM" src="https://user-images.githubusercontent.com/7317227/58827458-eceecc80-85f7-11e9-90c8-ff7b4eb898d1.png">

</details>

### Questions:
- Does it make sense to have these screens as part of the general onboarding wizard, or should these screens be a separate wizard since they only apply to sites with memberships (the first wizard in the Memberships checklist)?
- At what point should we tweak the WooCommerce settings for memberships sites? For example, disable coupons and product reviews. Currently that's done as part of adjusting the location settings, but that seems like a weird place to do that.
- @sonjaleix which Primary Button spacing do you prefer? Near the last field like the location settings screenshot, or with a little space between the button and the fields like the Stripe settings screenshot. I'll standardize it between the screens when I know which is best.

### How to test the changes in this Pull Request:

1. `npm run clean; npm run build`. **Make sure you have WooCommerce active**. We can adjust the logic once the wizard is fleshed out further to take into account sites that don't want to use WooCommerce, but the wizard wouldn't have anything to set up without WC so I'll do that in a future PR.
2. Navigate to Newspack > Onboarding in WP Admin, and go through the wizard.
3. This is the first wizard that will do a just-in-time install of a necessary plugin, so you can test that. Delete the Stripe plugin and toggle Stripe enabled on the wizard's Stripe settings. Stripe should be installed and activated when you click the 'Finish' button. If enabled is not toggled, Stripe should not be installed or activated.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

claudiulodro commented on assets/wizards/onboarding/views/locationSetup.js:
```diff
@@ -0,0 +1,97 @@
+					<SelectControl
+						label={ __( 'Where is your business based?' ) }
+						value={ countrystate }
+						options={ newspack_location_info }
```
Do we want to autocomplete the location and currency settings or something? Currently they are giant selects with 100+ options.

claudiulodro replied:
Not needed. See comment below.

jeffersonrabb commented:
> Make sure you have WooCommerce active. 

Any reason we're not also doing just-in-time installation of WooCommerce as well, or alternately use `PluginInstaller` to manage installing Woo?

jeffersonrabb commented:
> Do we want to autocomplete the location and currency settings or something? Currently they are giant selects with 100+ options.

I think we can call this a nice-to-have and leave for later. The  selects weren't so massive as to be unusable.

claudiulodro commented on includes/wizards/class-onboarding-wizard.php:
```diff
@@ -0,0 +1,339 @@
+			if ( $args['enabled'] ) {
+				// Stripe is not installed and we want to use it. Install/Activate/Initialize it.
+				Plugin_Manager::activate( 'woocommerce-gateway-stripe' );
+				do_action( 'plugins_loaded' );
```
I don't believe it always need to be called. It just needed to be called this time because the Stripe gateway hooks into that action to register the gateway with WooCommerce, and it needed to be registered with WC before we could use it. It's easy enough to refactor in the future if we find ourselves duplicating it a lot.

claudiulodro commented on includes/wizards/class-onboarding-wizard.php:
```diff
@@ -0,0 +1,339 @@
+	public function register_api_endpoints() {
+		// Get location info.
+		register_rest_route(
+			'newspack/v1/wizard/',
```
Good thinking. I've done that for this wizard. I'll do it for the Subscriptions Wizard in a separate PR, along with updating that wizard with PluginInstaller dependency management, error handling, etc. following the same pattern from this PR.

claudiulodro commented:
Excellent feedback @jeffersonrabb and @philipjohn!

I've implemented the following changes:
- [Renamed this to `Subscriptions Onboarding`](https://github.com/Automattic/newspack-plugin/pull/86/files#diff-4810642f3339991f021f7089c34f016fR26), with the intention to have this be the first wizard in the Memberships checklist. I don't think it should be the first couple screens of the existing Subscriptions wizard, as it would be a hassle for the user to get back to the subscription management screen if they need to go through a bunch of other screens first.
- [Handled the WooCommerce requirement with `PluginInstaller`](https://github.com/Automattic/newspack-plugin/pull/86/files#diff-07e22c86f90e246a09451508cbf0db5eR284) (that component wasn't available when I first opened this PR)
- [Added `newspack_string_to_bool` and `newspack_clean` utility functions](https://github.com/Automattic/newspack-plugin/pull/86/files#diff-2abf6b06afb600e6e28b06e6d84e3a58) (copied from WC), so we can use them for all our other wizards.
- [Added the wizard slug to the endpoint namespaces](https://github.com/Automattic/newspack-plugin/pull/86/files#diff-4810642f3339991f021f7089c34f016fR57), so we won't have conflicts between wizard endpoints.
- [Linked out to Stripe API keys info in the Stripe setup screen](https://github.com/Automattic/newspack-plugin/pull/86/files#diff-73361874a71863d6bd88c30daf1de5d3R81).
- Added error handling and form validation. There are two levels handled: notice and fatal. The main difference is that 'notice'-level errors can be fixed and retried; fatal can't. There are a couple stub features (e.g. "Return to checklist" doesn't do anything), and the validation errors could be more useful/displayed differently. I plan on improving those down the line after this PR is merged.

Notice-level error:

<img width="894" alt="Screen Shot 2019-06-07 at 10 17 03 AM" src="https://user-images.githubusercontent.com/7317227/59123242-a2c85c80-8911-11e9-8e83-8afd29417c54.png">

Fatal-level error:

<img width="644" alt="Screen Shot 2019-06-07 at 10 17 37 AM" src="https://user-images.githubusercontent.com/7317227/59123244-a5c34d00-8911-11e9-9e88-b59325557834.png">

### New things to test
- Deactivate/uninstall WooCommerce then go to the wizard. You should encounter a required plugin screen that can install WC. If WC is already installed, the required plugin screen should be automatically skipped.
- Try submitting the address form without some address info. A notice should appear and the submit should fail. After fixing the address info, the form should submit correctly.
- In a separate tab, deactivate WooCommerce halfway through the wizard. Try submitting the current wizard screen, and you should get a fatal-level notice about it.

### Next step after this is merged

I plan on evaluating similarities between the Subscriptions Onboarding wizard and the Subscriptions wizard and come up with a class, function, higher order component , etc., to reduce code duplication among the wizards and handle common functionality. I'll open a new PR for that after this is merged.

philipjohn commented:
Nice improvements :+1: I managed to find one quirk..

1. With WooCommerce deactivated/not installed, visit the wizard
2. Observe the checklist with "Use"/"Use All" buttons
3. Separately install/activate WooCommerce
4. Back in the checklist, click "Use" or "Use All"
5. The screen changes to the following, which gives no way forward for the user even though the conditions are met:

![Screenshot from 2019-06-10 12-46-58](https://user-images.githubusercontent.com/136342/59193921-3e123980-8b7f-11e9-87b1-6d33bd718896.png)

claudiulodro commented:
I've opened an issue around the plugin issues @philipjohn noticed: #97

> I'm able to get through the Stripe setup screen leaving either or both key blank. Shouldn't these be validated? There might be an argument that if these are going to be validated at all, this should be done by fully verifying that the keys work with Stripe—doing a simple "is it blank" validation might be insufficient or misleading. My suggestion would be to add non-blank validation for now, and file an issue for more sophisticated Stripe validation later.

Fixed in https://github.com/Automattic/newspack-plugin/pull/86/commits/e6abbdc0a672accddabba302c7ac04ff8f81acc1

> There's a bit of confusion around the Continue/Skip buttons. In a setup wizard like this where every step doesn't correspond directly to a Checklist Task, it's a little weird to have two buttons that both effectively do the same thing. This is one for @sonjaleix to consider, so we can establish a good pattern for all wizards.

The important difference between the two in my testing is that 'Skip' won't save the screen. This mean e.g. you don't need to be careful not to mess up locations settings when going to the Stripe settings. I do find myself using Skip quite a bit, but my use-case is not common since people won't be setting up Stripe 10 times. :)

I'm going to merge this PR then address the remaining issues in separate PRs.

sonjaleix commented:
@claudiulodro Just a few things we should discuss I noticed on first look: 
- Should we allow users to skip the first step "about your publication"
- Stripe setup: the link to get the API keys doesn't feel very intuitive since the full line is linked. I would recommend to keep it as a section header, and then add a link below that says "Learn how" or something like that.
- something minor: the tab in the left sidebar says "Subscription onboarding", we could probably just call it onboarding

Otherwise this is a great start for the onboarding wizard.

claudiulodro commented:
Good points @sonjaleix! I agree with your logic, and I'll open a new PR with the first 2 changes. The third one shouldn't be an issue since the [wizards don't show up in the sidebar](https://user-images.githubusercontent.com/7317227/59218292-c2ab8a80-8b74-11e9-947e-64a49b2df311.png) after #96.
